tests/tt_metal/tt_metal/unit_tests/tt_stl/test_any_range.cpp

```diff
@@ -0,0 +1,150 @@
+
+TEST(AnyRangeTest, CanTypeEraseSizedRandomAccessRange) {
+    AnySizedRandomAccessIntRange range{std::vector{1, 2, 3, 4, 5}};
+
+    auto begin = range.begin();
+    const auto end = range.end();
+    begin[2] = 9;
+
+    EXPECT_EQ(*begin, 1);
+    ++begin;
+    EXPECT_EQ(*begin, 2);
+    ++begin;
+    EXPECT_EQ(*begin, 9);
+    ++begin;
+    EXPECT_EQ(*begin, 4);
+    ++begin;
+    EXPECT_EQ(*begin, 5);
+    ++begin;
+    EXPECT_EQ(begin, end);
+
+    {
+        AnySizedRandomAccessIntRange another_range{std::deque{6, 7, 8}};
+
+        using std::swap;
+        swap(range, another_range);
+
+        // another_range is now bound to std::vector from range
+        const auto begin = another_range.begin();
+        const auto end = another_range.end();
+
+        EXPECT_EQ(end - begin, 5);
+        EXPECT_EQ(end[-1], 5);
+        EXPECT_EQ(end[-2], 4);
+        EXPECT_EQ(end[-3], 9);
+        EXPECT_EQ(end[-4], 2);
+        EXPECT_EQ(end[-5], 1);
+        EXPECT_EQ(end - 5, begin);
+    }
+
+    // range is now bound to std::deque from another_range
+    auto rbegin = range.rbegin();
+    const auto rend = range.rend();
+
+    EXPECT_EQ(rbegin[0], 8);
+    EXPECT_EQ(rbegin[1], 7);
+    EXPECT_EQ(rbegin[2], 6);
+    EXPECT_EQ(rbegin + 3, rend);
+    EXPECT_EQ(rend - rbegin, 3);
+    EXPECT_EQ(range.size(), 3);
+    EXPECT_FALSE(range.empty());
+
+    range[1] = 9;
+    std::sort(range.rbegin(), range.rend());
+
+    EXPECT_EQ(range[0], 9);
+    EXPECT_EQ(range[1], 8);
+    EXPECT_EQ(range[2], 6);
+
+    {
+        std::vector vector{10, 11, 12, 13, 14, 15};
+        // a non-owning type-erased range, it stores vector as a reference
+        range = vector;
+
+        // modifying vector is observable through range
+        vector[0] = 17;
+        EXPECT_EQ(range[0], 17);
+
+        auto begin = range.begin();
+        auto end = range.end();
+
+        // non-owning range writes back to vector
+        begin += 3;
+        *begin = 20;
+
+        EXPECT_EQ(vector[3], 20);
+
+        EXPECT_FALSE(begin == end);
+        EXPECT_TRUE(begin != end);
+        EXPECT_TRUE(begin < end);
+        EXPECT_FALSE(begin > end);
+        EXPECT_TRUE(begin <= end);
+        EXPECT_FALSE(begin >= end);
+
+        end -= 3;
+
+        EXPECT_EQ(end - begin, 0);
+
+        EXPECT_TRUE(end == begin);
+        EXPECT_FALSE(end != begin);
+        EXPECT_FALSE(end < begin);
+        EXPECT_FALSE(end > begin);
+        EXPECT_TRUE(end <= begin);
+        EXPECT_TRUE(end >= begin);
+
+        EXPECT_TRUE(vector.begin() == range.begin());
+        EXPECT_TRUE(vector.end() == range.end());
+        EXPECT_TRUE(vector.rbegin() == range.rbegin());
+        EXPECT_TRUE(vector.rend() == range.rend());
+    }
+
+    // since range is non-owning, usage here would be UB after the vector has ended its lifetime
+
+    {
+        std::deque deque{10, 11, 12, 13, 14, 15};
+        // an owning type-erased deque, it stores deque as a value
+        range = std::move(deque);
```

Comment:
Is my understanding correct that `range = std::move(deque)` takes ownership, but `range = deque` would just store a reference to deque? It seems it would be extremely easy to do the wrong thing by accident, can we consider doing a named explicit method here to prevent such confusion?

Reply:
I can add named methods that delegate to this, but I would prefer to keep the forwarding constructor I currently have in place, as it makes the syntax reasonably succinct.

Reply:
It seems very scary and error-prone to me.
For example `auto new_range = range;` would create a non-owning reference, but the user who got an owning range, might've expected a copy. This seems like one of the bugs which is extremely easy to make, hard to debug, and impossible to spot in the code review.

Reply:
`auto new_range = range;` does not create a non-owning reference, it selects the copy constructor of the AnyRange which always does the correct thing.